webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/BinaryDataService.scala

```diff
@@ -199,19 +204,34 @@ class BinaryDataService(val dataBaseDir: Path,
     compressed
   }
 
-  def clearCache(organizationName: String, dataSetName: String, layerName: Option[String]): (Int, Int) = {
+  def clearCache(organizationName: String, datasetName: String, layerName: Option[String]): (Int, Int, Int) = {
+    val dataSourceId = DataSourceId(datasetName, organizationName)
+
     def dataCubeMatchPredicate(cubeKey: CachedCube) =
-      cubeKey.dataSourceName == dataSetName && cubeKey.organization == organizationName && layerName.forall(
+      cubeKey.dataSourceName == datasetName && cubeKey.organization == organizationName && layerName.forall(
         _ == cubeKey.dataLayerName)
 
     def agglomerateFileMatchPredicate(agglomerateKey: AgglomerateFileKey) =
-      agglomerateKey.dataSetName == dataSetName && agglomerateKey.organizationName == organizationName && layerName
+      agglomerateKey.dataSetName == datasetName && agglomerateKey.organizationName == organizationName && layerName
         .forall(_ == agglomerateKey.layerName)
 
+    def bucketProviderPredicate(key: (DataSourceId, String)): Boolean =
+      key._1 == DataSourceId(datasetName, organizationName) && layerName.forall(_ == key._2)
+
     val closedAgglomerateFileHandleCount =
       agglomerateServiceOpt.map(_.agglomerateFileCache.clear(agglomerateFileMatchPredicate)).getOrElse(0)
+
     val closedDataCubeHandleCount = shardHandleCache.clear(dataCubeMatchPredicate)
-    (closedAgglomerateFileHandleCount, closedDataCubeHandleCount)
+
+    bucketProviderCache.clear(bucketProviderPredicate)
+
+    def chunkContentsPredicate(key: String): Boolean =
```

Comment:
Not totally happy with this being placed here, since the definition of the key is somewhere else. Maybe move this predicate to where the key is defined and only use it here?

Reply:
Hmm, yes I see what you mean. However, the same would be the case for the other two predicate functions here. So we could either extract each of those so that there is only one place for each cache, where cache key building happens. This would lead to less distributed knowledge on how the keys work. However, I also like all of these methods being together here, since it is less distributed knowledge about the fact that these caches should be cleared dataset-wise or layer-wise. So it’s some fragmentation either way. If you’re ok with it, I’d leave it as it is for the moment.